File: 240322/두 방향 탈출 가능 여부 판별하기/determine-escapableness-with-2-ways.cpp
#include <iostream>
using namespace std;
#define MAX_NM 100

int n, m;
int grid[MAX_NM][MAX_NM];
int visited[MAX_NM][MAX_NM];
int dx[2] = {1, 0}, dy[2] = {0, 1};

bool InRange(int x, int y) {
    return 0<=x && x<n && 0<=y && y<n;
}

bool CanGo(int x, int y) {
    if(!InRange(x, y)) return false;

    if(grid[x][y] == 0 || visited[x][y] == 1) return false;

    return true;
}

void DFS(int vertex_x, int vertex_y) {
    for(int i=0; i<2; i++) {
        int nx = vertex_x+dx[i], ny = vertex_y+dy[i];
        if(CanGo(nx, ny)) {
            visited[nx][ny] = 1;
            DFS(nx, ny);
        }
    }

}

int main() {
    // 여기에 코드를 작성해주세요.
    cin >> n>>m;
    for(int i=0; i<n; i++) {
        for(int j=0; j<m; j++) {
            cin >> grid[i][j];
        }
    }

    visited[0][0] = 1;
    DFS(0, 0);

    if(visited[0][0] && visited[n-1][n-1]) cout << 1;
    else cout << 0;

    return 0;
}

File: 250128/두 방향 탈출 가능 여부 판별하기/determine-escapableness-with-2-ways.py
n, m = tuple(map(int, input().split()))

grid = [
    list(map(int, input().split()))
    for _ in range(n)
]

visited = [[0 for _ in range(m)] for _ in range(n)]

dxs, dys = [0, 1], [1, 0]

def inRange(x, y):
    return 0<=x<n and 0<=y<n

def canGo(x, y):
    if not inRange(x, y):
        return False
    if visited[x][y] or grid[x][y] == 0:
        return False
    
    return True

def dfs(x, y):
    for dx, dy in zip(dxs, dys):
        nx, ny = x+dx, y+dy

        if canGo(nx, ny):
            visited[nx][ny] = 1
            dfs(nx, ny)


visited[0][0] = 1
dfs(0, 0)

print(visited[n-1][m-1])